src/Exceptionless.Core/Services/StackService.cs

```diff
@@ -0,0 +1,94 @@
+            if (String.IsNullOrEmpty(organizationId) || String.IsNullOrEmpty(projectId) || String.IsNullOrEmpty(stackId) || count == 0)
+                return;
+            var expireTimeout = TimeSpan.FromDays(1);
+            var tasks = new List<Task>(4);
+
+            string occurenceCountCacheKey = GetStackOccurrenceCountCacheKey(organizationId, projectId, stackId),
+                occurrenceMinDateCacheKey = GetStackOccurrenceMinDateCacheKey(organizationId, projectId, stackId),
+                occurrenceMaxDateCacheKey = GetStackOccurrenceMaxDateCacheKey(organizationId, projectId, stackId),
+                occurrenceSetCacheKey = GetStackOccurrenceSetCacheKey();
+
+            var cachedOccurrenceMinDateUtc = await _cache.GetAsync<DateTime>(occurrenceMinDateCacheKey).AnyContext();
+            if (!cachedOccurrenceMinDateUtc.HasValue || cachedOccurrenceMinDateUtc.IsNull || cachedOccurrenceMinDateUtc.Value > minOccurrenceDateUtc)
+                tasks.Add(_cache.SetAsync(occurrenceMinDateCacheKey, minOccurrenceDateUtc, expireTimeout));
+
+            var cachedOccurrenceMaxDateUtc = await _cache.GetAsync<DateTime>(occurrenceMaxDateCacheKey).AnyContext();
+            if (!cachedOccurrenceMaxDateUtc.HasValue || cachedOccurrenceMinDateUtc.IsNull || cachedOccurrenceMaxDateUtc.Value < maxOccurrenceDateUtc)
+                tasks.Add(_cache.SetAsync(occurrenceMaxDateCacheKey, maxOccurrenceDateUtc, expireTimeout));
+
+            tasks.Add(_cache.IncrementAsync(occurenceCountCacheKey, count, expireTimeout));
+            tasks.Add(_cache.SetAddAsync(occurrenceSetCacheKey, Tuple.Create(organizationId, projectId, stackId), expireTimeout));
+
+            await Task.WhenAll(tasks).AnyContext();
+        }
+
+        public async Task SaveStackUsagesAsync(bool sendNotifications = true, CancellationToken cancellationToken = default) {
+            var occurrenceSetCacheKey = GetStackOccurrenceSetCacheKey();
+            var stackUsageSet = await _cache.GetSetAsync<Tuple<string, string, string>>(occurrenceSetCacheKey).AnyContext();
+            if (!stackUsageSet.HasValue || stackUsageSet.IsNull) return;
+            foreach (var tuple in stackUsageSet.Value) {
+                if (cancellationToken.IsCancellationRequested) break;
+
+                string organizationId = tuple.Item1, projectId = tuple.Item2, stackId = tuple.Item3;
+                string occurenceCountCacheKey = GetStackOccurrenceCountCacheKey(organizationId, projectId, stackId),
+                    occurrenceMinDateCacheKey = GetStackOccurrenceMinDateCacheKey(organizationId, projectId, stackId),
+                    occurrenceMaxDateCacheKey = GetStackOccurrenceMaxDateCacheKey(organizationId, projectId, stackId);
+                var occurrenceCount = await _cache.GetAsync<long>(occurenceCountCacheKey, 0).AnyContext();
```

Comment:
I wish we could do a get and decrement at the same time (atomically), I feel like this could be a race condition long term (but will be eventually consistent)

Reply:
When saving stack usages, we will retrieve the count of the events in the period between last saving and this saving and increment this number to elasticsearch, then we should remove the occurrence count or set it 0 for the incoming event post job processing.  
Suppose we want to achieve this purpose through the `decrement` operation, we have to know the occurrence count before operation.
I think the better choice is: Add new method for the `ICacheClient` that can get and set cache value at the same time atomically. As I know there is API for this operation
https://github.com/StackExchange/StackExchange.Redis/blob/c4c9c1fdb455070415e82d2f104fc89a90b057b5/StackExchange.Redis/StackExchange/Redis/IDatabase.cs#L1020
and
https://github.com/StackExchange/StackExchange.Redis/blob/c4c9c1fdb455070415e82d2f104fc89a90b057b5/StackExchange.Redis/StackExchange/Redis/IDatabaseAsync.cs#L960